src/transaction.ts

```diff
@@ -446,9 +571,31 @@ class Transaction extends DatastoreRequest {
       typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
     const callback =
       typeof optionsOrCallback === 'function' ? optionsOrCallback : cb!;
-    this.#runAsync(options).then((response: RequestPromiseReturnType) => {
-      this.#processBeginResults(response, callback);
-    });
+    const runIfStarted = () => {
+      process.emitWarning(
+        'run has already been called and should not be called again.'
+      );
+      callback(null, this, {transaction: this.id});
+    };
+    if (this.#state === TransactionState.NOT_STARTED) {
+      this.#mutex.acquire().then(release => {
+        if (this.#state === TransactionState.NOT_STARTED) {
+          this.#runAsync(options).then(
+            (
+              response: UserCallbackData<google.datastore.v1.IBeginTransactionResponse>
+            ) => {
+              release();
+              this.#processBeginResults(response, callback);
```

Comment:
https://github.com/googleapis/nodejs-datastore/pull/1205/files#diff-7aa7c28abfaf34d94e1f9755555bd2571bad57fb652aad577c041cd5adf630c6R1018 is a test that captures this case and ensures that the `beginTransaction` call only happens once. The placement of release doesn't matter because the release is an async event that gets added to the end of the event loop.

In any case, I replaced this with `runExclusive` so that we don't have to trigger the release programmatically.